(H)omotopy CGM

gideonite opened the pull request:
This is a first step towards implementing Homotopy CGMs into copt.

Based closely on the existing Frank-Wolfe implementation, this
incorporates additional linear constraints into the picture via homotopy
smoothing.

In this commit there are element-wise and row-wise constraints.

The next step is to provide randomized versions of this algorithm.

gideonite committed:
(H)omotopy CGM

This is a first step towards implementing Homotopy CGMs into copt.

Based closely on the existing Frank-Wolfe implementation, this
incorporates additional linear constraints into the picture via homotopy
smoothing.

In this commit there are element-wise and row-wise constraints.

The next step is to provide randomized versions of this algorithm.

gideonite commented:
I should add some unit tests for this as well, but I thought that now might be a good time to get feedback.

gideonite commented:
Note to self: are there places where I could make the code faster? Specifically, but using numba / utils.njit? There aren't any tight loops where I am going over an array. The basic Frank-Wolfe implementation doesn't have numba stuff so maybe we can't either, but I can help thinking that there might be a way to accelerated the main for-loop of the algorithm.

fabianp commented on copt/constraint.py:
```diff
@@ -278,14 +278,16 @@ class TraceBall:
         alpha: float
             radius of the ball.
 
+        use_eigs: use an eigenvalue solver rather than an SVD solver. This
```
Is this necessary? under the hood svds is using the same eigensolver you'r using below (well, almost, it's using the Hermitian version eigsh) (https://github.com/scipy/scipy/blob/v1.6.2/scipy/sparse/linalg/eigen/arpack/arpack.py#L1727-L1913) 

If the issue is that the tolerance is too low, I think we should use the tolerance parameter in svds instead

gideonite replied:
Interesting. Maybe the tolerance was the issue! I found that they gave quite different solutions, with eigs reproducing what was in their paper and svds not. I was also thinking that it was strange though. Hopefully, it's the tolerance.

fabianp commented on copt/loss.py:
```diff
@@ -296,3 +296,32 @@ def f_grad(self, x, return_gradient=True):
         return (s * s) / self.A.shape[0] + self.alpha
+
+
+class LinearObjective:
```
can you rename to LinearLoss for consistency with the rest of losses?

fabianp commented on examples/homotopy/plot_homotopy.py:
```diff
@@ -0,0 +1,106 @@
+    n_labels = len(np.unique(labels))
+    return C, n_labels, opt_val
+
+class HomotopyTrace(copt.utils.Trace):
```
please describe (missing docstring)

fabianp commented on copt/constraint.py:
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
         return vertex - x, None, None, 1.
+
+
+class RowEqualityConstraint:
```
please describe (missing docstring)

fabianp commented on copt/constraint.py:
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
+        return val / self.offset_norm
+
+
+class ElementWiseInequalityConstraint:
```
please describe (missing docstring)

fabianp commented on copt/constraint.py:
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
+        z = np.matmul(X, self.operator)
+        return np.all(z == self.offset)
+
+    def smoothed_grad(self, x):
```
please describe (missing docstring)

fabianp commented on copt/constraint.py:
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
+        grad = 2*np.outer(err, self.operator)
+        return val, grad.flatten()
+
+    def feasibility(self, x):
```
please describe (missing docstring)

fabianp commented on copt/constraint.py:
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
+    def __call__(self, x):
+        return np.all(x >= self.offset)
+
+    def smoothed_grad(self, x):
```
please describe (missing docstring)

fabianp commented on examples/homotopy/plot_homotopy.py:
```diff
@@ -0,0 +1,106 @@
+
+basepath = Path(".")
+def reduced_digits():
+    mat = sio.loadmat(basepath / "data/reduced_clustering_mnist_digits.mat")
```
It would be nice if this dataset was generally available from the library (https://github.com/openopt/copt/blob/master/copt/datasets.py) and not just specific for this example. This way, we could re-use it easier from other examples

fabianp reviewed (commented):
thanks for the contribution! Made a quick pass and left some early comments

gideonite commented:
Super, thanks you! I'll do these things and update.

gideonite committed:
some doc strs

gideonite committed:
docs and error fix

gideonite committed:
docs and error fix in main loop

gideonite committed:
revert traceball to master

gideonite committed:
spectrahedron constraint

gideonite committed:
move symmetrization into spectrahedron

gideonite committed:
rename, doc str, use spectrahedron

gideonite committed:
doc str

gideonite committed:
refactor dataset loading

still need to handle downloading

gideonite committed:
dont commit datasets

gideonite commented:
Rather than this clumsy `use_eigs` option that was there before, I replaced it with a new lmo, TraceSpectrahedron, which optimizes over the psd matrices with bounded trace norm. This is actually what I originally wanted but I was a bit confused.

Other than that, I added the comments, changed the names, etc. as you requested. @fabianp 

Still to do:
1. [unit tests] I don't need feedback on this (yet). I just need to do it.
2. [figure out how dataset downloading should work] I'm not sure how to proceed or whether you have an opinion. I can add the dataset to the google storage but I need access. Or, I download directly from the source. The task is clustering so I'm not sure how much of the `_load_dataset` function should be reused. @fabianp

fabianp commented:
No worries about the history, it's fine if there are multiple commits

gideonite committed:
trace spectrahedron unit test & cleanup

fabianp commented:
another option for the dataset is to use something more standard that already includes mnist like tf-datasets, https://www.tensorflow.org/datasets/overview

gideonite committed:
test non-neg constraint

gideonite committed:
test for row constraint

gideonite committed:
remove dead assertion

gideonite committed:
remove ignored parameters

gideonite committed:
download data

gideonite commented:
The only problem with that is that the dataset that we are using are the outputs of a softmax. We could train the simple model dynamically when the dataset is downloaded or download directly what was used in [MVW16]. For now, I decided to download it directly.

Ok, I think it's finally ready for merge! What do you think? @fabianp 

[MVW16] D. G. Mixon, S. Villar, and R. Ward, “Clustering subgaussian mixtures by semidefinite programming,” May 2016. <http://arxiv.org/abs/1602.06612>, GitHub Repo: <https://github.com/solevillar/kmeans_sdp>

gideonite committed:
fix symmetrization

gideonite committed:
basically revert to non-negative constraint

gideonite committed:
fix sdp mnist dataset preprocessing

gideonite committed:
some kind of unit test for homotopy

gideonite commented:
Fixed a small bug and added a unit test to test for it next time. Apparently, this unit test also improves the code coverage above the threshold for coveralls (finally).

fabianp commented on copt/homotopy.py:
```diff
@@ -0,0 +1,102 @@
+
+EPS = np.finfo(np.float32).eps
+
+# TODO :ref:`hcgm`
```
Please remove TODO or at least make it clearer

gideonite replied:
Whoops, sorry about that. Removed.

fabianp commented:
thanks @gideonite l it all looks good to me.

@GeoffNN do you mind taking a look a this ? Thanks

GeoffNN commented:
I'll look at it ASAP!

gideonite committed:
remove completed todo and minor formatting

gideonite committed:
dont need the trace

gideonite commented:
Does anyone else have feedback? This has been in purgatory for a while and I'd like to get it merged. @GeoffNN @fabianp

GeoffNN reviewed (commented):
A few cosmetic comments, the rest looks good to me :) Will approve when you answer @gideonite, and sorry for the wait!!

GeoffNN commented:
@gideonite quick ping on this, if you want to close it up!

gideonite commented:
Hi @GeoffNN, sorry for the delay. Been caught up with other things and this has fallen by the wayside. I will need some more time to get back to this but if there's still interest in getting this merged, I can probably easily address most of these issues.

GeoffNN commented:
The remaining issues should be straightforward / it would also be fine to leave TODO items for a layer push :)

gideonite commented:
I think it's not happening :-)
Thanks for the support

fabianp commented:
LOL
